glp/neighborlist.py

```diff
@@ -91,11 +105,23 @@ def cell_too_small(new_cell):
         )
         comms.warn("this will yield incorrect results!")
 
-    squared_cutoff = (cutoff + skin) ** cast(2.0)
+    cutoff = (cutoff + skin)
 
     allowed_movement = (skin * cast(0.5)) ** cast(2.0)
 
+    if cell is not None and use_cell_list:
+        if jnp.all(jnp.any(cutoff < get_heights(cell) / 3., axis=0)):
+            cl_allocate, cl_update = cell_list(cell, cutoff)
+        else:
+            cl_allocate = cl_update = None
+    else:
+        cl_allocate = cl_update = None
+        
     def need_update_fn(neighbors, new_positions, new_cell):
```

Comment:
In this case you are right, but is not this lines of the code.
``` python
    if cell is not None and use_cell_list:
        if jnp.all(jnp.any(cutoff < get_heights(cell) / 3., axis=0)):
            cl_allocate, cl_update = cell_list(cell, cutoff)
        else:
            cl_allocate = cl_update = None
    else:
        cl_allocate = cl_update = None
```
This block is essentially checking if it’s possible to divide the cell into at least 3 bins along any of the cell vectors. If it’s not possible, it doesn’t make sense to use the cell list. It might be beneficial to use a cutoff plus some tolerance. The need for reallocation of the cell list arises not when the cell gets larger, but when it gets smaller such that the bin is not large enough to guarantee that it can search for each neighbor (i.e., the height in one or more directions is less than the cutoff).

``` python
    def update_fn(positions, old_cell_list, new_cell):
        # this is jittable,
        # CellList tells us all the shapes we need
        # If bin size is lower than cutoff, we need to reallocate cell list
        # Rellocate is not jitable, we're changing shapes
        # So, after each update we need to check if we need to reallocate
        N = positions.shape[0]

        bin_size = jnp.where(old_cell_list.bins_per_side != 0, (cell.T / old_cell_list.bins_per_side).T, cell)
        max_occupancy = estimate_bin_capacity(positions, new_cell, bin_size, 1)
        # Checking if update or reallocate
        reallocate = jnp.all(get_heights(bin_size).any() >= cutoff) & (max_occupancy <= old_cell_list.capacity)
```
Here, the code updates the `bin_size` according to the new cell keeping number of bins per side. Afterwards, it checks the occupancy (if there are more atoms in one bin than bin_capacity) and if the height is equal to or greater than the cutoff.